include/dmlc/threadediter.h

```diff
@@ -278,120 +295,181 @@ Init(Producer *producer, bool pass_ownership) {
-            free_cells_.push(queue_.front());
-            queue_.pop();
+        }  // end of lock scope
+        // now without lock
+        produce_end_ = !next(&cell);
+        DCHECK(cell != NULL || produce_end_);
+        bool notify;
+        {
+          // lockscope
+          std::lock_guard<std::mutex> lock(mutex_);
+          if (!produce_end_) {
+            queue_.push(cell);
+          } else {
+            if (cell != NULL)
+              free_cells_.push(cell);
           }
-          // reset the state
-          produce_end_ = false;
-          producer_sig_processed_ = true;
-          producer_sig_ = kProduce;
-          // notify consumer that all the process as been done.
-          lock.unlock();
-          consumer_cond_.notify_all();
-          continue;
-        } else {
-          // destroy the thread
-          CHECK(producer_sig_ == kDestroy);
-          producer_sig_processed_ = true;
-          produce_end_ = true;
+          // put things into queue
+          notify = nwait_consumer_ != 0;
+        }
+        if (notify)
           consumer_cond_.notify_all();
-          return;
+      } catch (dmlc::Error &e) {
+        // Shouldn't throw exception in destructor
+        DCHECK(producer_sig_ != kDestroy);
+        {
+          std::lock_guard<std::mutex> lock(mutex_exception_);
+          if (!iter_exception_) {
+            iter_exception_ = std::current_exception();
+          }
         }
-      }  // end of lock scope
-      // now without lock
-      produce_end_ = !next(&cell);
-      CHECK(cell != NULL || produce_end_);
-      bool notify;
-      {
-        // lockscope
-        std::lock_guard<std::mutex> lock(mutex_);
-        if (!produce_end_) {
-          queue_.push(cell);
-        } else {
-          if (cell != NULL) free_cells_.push(cell);
+        bool next_notify = false;
+        {
+          std::unique_lock<std::mutex> lock(mutex_);
+          if (producer_sig_ == kBeforeFirst) {
+            while (queue_.size() != 0) {
+              free_cells_.push(queue_.front());
+              queue_.pop();
+            }
+            produce_end_ = true;
+            producer_sig_processed_ = true;
+            lock.unlock();
+            consumer_cond_.notify_all();
+          } else if (producer_sig_ == kProduce) {
+            produce_end_ = true;
+            next_notify = nwait_consumer_ != 0;
+            lock.unlock();
+            if (next_notify)
+              consumer_cond_.notify_all();
+          }
         }
-        // put things into queue
-        notify = nwait_consumer_ != 0;
+        return;
       }
-      if (notify) consumer_cond_.notify_all();
     }
   };
   producer_thread_ = new std::thread(producer_fun);
 }
 
-template<typename DType>
-inline bool ThreadedIter<DType>::
-Next(DType **out_dptr) {
-  if (producer_sig_ == kDestroy) return false;
+template <typename DType>
+inline bool ThreadedIter<DType>::Next(DType **out_dptr) {
+  if (producer_sig_ == kDestroy)
```

Comment:
why would this be called during destruction?

Reply:
this was not added by my code.

Reply:
consumer thread may not notice there is destruction happening. The producer thread need to signal destruction and consumer receives the signal and then exit

Reply:
@tqchen Thank you for the explanation! Since there is only a single consumer and the Destroy method is called by the consumer, which results in signal to producer and then notifying Consumer. I am not sure how Next is being called here.  "The producer thread need to signal destruction and consumer receives the signal and then exit". Can you please elaborate on this.  Its not waiting for any signal on this line : `if (producer_sig_ == kDestroy)`

Reply:
OK, I think that was a behavior which is not necessary. As it is always the consumer who calls the destroy. The check is just a safeguard for undesired usecase

Reply:
thanks for the clarification!